sdks/python/apache_beam/io/fileio.py

```diff
@@ -708,10 +708,26 @@ def write(self, data):
 
   def _fetch_to_internal_buffer(self, num_bytes):
     """Fetch up to num_bytes into the internal buffer."""
-    while not self._read_eof and len(self._data) < num_bytes:
+    if (not self._read_eof and self._read_position > 0 and
+        (self._read_buffer.tell() - self._read_position) < num_bytes):
+      # There aren't enough number of bytes to accommodate a read, so we
+      # prepare for a possibly large read by clearing up all internal buffers
+      # but without dropping any previous held data.
+      self._read_buffer.seek(self._read_position)
+      data = self._read_buffer.read()
+      self._read_position = 0
+      self._read_buffer.truncate(0)
```

Comment:
I'm not sure if truncate does this (it doesn't for ordinary files), so probably we should.